data/idol.1
.\" Copyright (C) 2004 Andre Lehovich <andrel@u.arizona.edu>
.\"
.\" This is free software; you may redistribute it and/or modify
.\" it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2,
.\" or (at your option) any later version.
.\"
.\" This is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License 
.\" along with this; if not write to the Free Software Foundation, Inc.
.\" 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA
.TH idol 1 "2008\-08\-25" "MATE"
.SH NAME
Idol \- MATE desktop movie player based on GStreamer
.SH SYNOPSIS
.B idol
.RI [ option... ] " " [ filenames | URI ]
.SH DESCRIPTION
.B Idol
is a movie player for the MATE desktop.  Through use of the
.B GStreamer
backend many popular file and streaming formats are supported including:
MPEG audio and video (MP3/MPV), Ogg, AVI, ASF, MOV, WMV,
QuickTime, and more.  Playing of DVD and VCD media is
also possible.
Although primarily intended as a movie player,
.B Idol
can also play many audio formats.
.P
For additional documentation on interactive use of 
.B Idol
select the
.I Help
menu option.  The menus allow access to some features not
available through keybindings or command-line options.
.SH OPTIONS
.TP
.B \-?, --help
Output a longer help message then quit.  (Any discrepancies
between this man page and the output of
.B idol --help
should be reported as a documentation bug.)
.TP
.B --version
Output version information then quit.
.P
The following options command an already-running instance of
.B Idol
to do something; they are useful for remote-control scripting. If 
.B Idol
is not already running, these commands will launch a new instance of
.B Idol
but will not do anything further.
.TP
.B --play-pause
Tell an already-running instance of 
.B Idol
to toggle between play and pause.
.TP
.B --play
Tell an already-running instance of
.B Idol
to play (has no effect if already playing)
.TP
.B --pause
Tell an already-running instance of
.B Idol
to pause (has no effect if already paused)
.TP
.B --next
Tell an already-running instance of 
.B Idol
to skip to the next movie or chapter in the playlist.
.TP
.B --previous
Tell an already-running instance of 
.B Idol
to return to the previous movie or chapter in the playlist.
.TP
.B --seek-fwd
Tell an already-running instance of 
.B Idol
to seek forwards 15 seconds in the current movie.
.TP
.B --seek-bwd
Tell an already-running instance of 
.B Idol
to seek backwards 15 seconds in the current movie.
.TP
.B --volume-up
Tell an already-running instance of 
.B Idol
to raise the volume by 8%.
.TP
.B --volume-down
Tell an already-running instance of 
.B Idol
to lower the volume by 8%.
.TP
.B --fullscreen
Tell an already-running instance of 
.B Idol
to toggle fullscreen mode.
.TP
.B --toggle-controls
Tell an already-running instance of
.B Idol
to toggle showing the controls.
.TP
.B --quit
Tell an already-running instance of
.B Idol
to quit.
.TP
.BI "--enqueue " filename|URI
Tell an already-running instance of Idol to add a new stream
to the playlist.
.TP
.BI "--replace " filename|URI
Tell an already-running instance of Idol to play 
from the playlist.
.SH KEYBINDINGS
Global keybindings for
.BR Idol :
.P
.ta \w'Down-arrow   'u	
a	cycle between aspect ratios
.br
p	toggle between play and pause
.br
Esc	exit full screen mode
.br
f	toggle full screen
.br
h	toggle display of on-screen controls
.br
0	resize window to 50% original size
.br
1	resize window to 100% original size
.br
2	resize window to 200% original size
.br
.br
r	zoom in the video
.br
t	zoom out the video
.br
d	start and stop the telestrator (drawing) mode
.br
e	erase the drawing
.br
Left-arrow        skip back 15 seconds
.br
Right-arrow       skip forward 60 seconds
.br
Shift+Left-arrow  skip back 5 seconds
.br
Shift+Right-arrow skip forward 15 seconds
.br
Ctrl+Left-arrow   skip back 3 minutes
.br
Ctrl+Right arrow  skip forward 10 minutes
.br
Up-arrow	increase volume by 8%
.br
Down-arrow	decrease volume by 8%
.br
b		jump back to previous chapter/movie in playlist
.br
n		jump to next chapter/movie in playlist
.br
q		quit
.br
Ctrl+E          eject the playing optical media
.br
Ctrl+O		open a new file
.br
Ctrl+L		open a new URI
.br
F9		toggle display of the playlist
.br
m		show the DVD menu
.br
c		show the DVD chapter menu
.SH FILES
.I /usr/lib/idol/plugins/
.RS
The global plugins directory, which contains all the system-installed
plugins.
.RE
.I ~/.local/share/idol/plugins/
.RS
The user plugins directory, containing plugins available only to that user.
It is always in the "idol/plugins" subdirectory of your 
.B XDG_DATA_HOME
directory.
.RE
.I ~/.config/idol/state.ini
.RS
The configuration file for the main window's appearance, stored in a
simple INI format.
.RE
.SH BUGS
.B Idol
cannot play files using some proprietary codecs.
.SH AUTHOR
.B Idol
was written by Bastien Nocera <hadess@hadess.net>.
This manual page was written by Andre Lehovich for the
Debian Project.
.SH DISTRIBUTION
The latest version of Idol may be downloaded from
.UR http://projects.mate.org/idol/
<http://projects.mate.org/idol/>
.UE
.SH SEE ALSO
.BR "idol-video-thumbnailer" (1),
.P
The online documentation available through the program's
.I Help
menu.


NEWS
New features and significant updates in version...

2.32.0
 * Update translations

2.31.92.2
 * Port back from GApplication to libunique

2.31.92.1
* Fix a couple more uses of deprecated GDK API

2.31.92
* Stop using deprecated GDK API
* Some minor plugin documentation fixes
* Updated translations

2.31.90
* Add chapters plugin
* Support Web-M videos in YouTube plugin
* Fix possible crash when quitting whilst the stream's tags are being
  gathered
* Make it possible to use --raw along with --size in thumbnailer
* Loads of updated translations

2.31.6:
* Add support for Web-M to browser, movie player, and YouTube plugin

* Movie Player:
  - Add deinterlacing support, automatically used for interlaced content
  - Do not block when loading playlists
  - Remove libunique dependency
  - Make auto-resize use half-scale or 2x if the player would
    fit on the screen
  - Make screensaver preference clearer
  - Support running with RGBA (semi-transparent) GTK+ themes

* Plugins:
  - Update for YouTube website changes

* Browser plugin:
  - Fix running browser plugin in non-GTK+ front-ends

* VLC-compatible browser plugin:
  - Support input.time and input.length property
  - Support togglePause function
  - Support toolbar plugin argument

2.30.2:
* Movie Player:
  - Fix huge window height when hiding then showing controls again
  - Fix slight resize when showing the sidebar for the first time
  - Fix possible crash when loading PytGTK
  - Fix loading OSD icons when there's no SVG versions

* Browser Plugin:
  - Fix volume being at zero when starting playback
  - Fix use of URL parameter for the Window Media Player
    compatibility plugin
  - Fix position on End-Of-Stream when using the QuickTime
    compatibility plugin

* GStreamer backend:
  - Try not to block the UI when seeking
  - Fix constant buffering when playing live streams

2.30.1:
* Movie Player:
  - Fix artifacts and flashing or black screen when paused
  - Better download buffering handling
  - Fix crasher when PyGTK is badly installed
  - Avoid problems with seeking when pausing the video too early
  - Fix link to Idol website in error messages
  - Disable seeking in DVD menus
  - Make Eject menu item work for DVDs
  - Fix playback from FTP sites

* Plugins:
  - Port Tracker plugin to Tracker 0.8
  - Make YouTube plugin work again for latest site changes

* Browser Plugin:
  - Add download buffering support even when streaming from
    the browser itself
  - Fix playback on the Apple trailers site, and add seeking support

2.30.0:
* Updated translations

2.29.92:
* Fix some possible hangs when switching files

2.29.91:
* Fix compilation with newer versions of Tracker
* Update time in the status bar when frame-stepping
* Fix a lot of build problems with newer GTK+, or using pedantic linkers
* Update Vala bindings for external plugins
* Fix problems with the popup menu in the browser plugin

2.29.4:
* Fix documentation build

2.29.3:
* Movie Player:
  - Add MateConf key to disable all the keyboard shortcuts
  - Implement type-ahead search
  - Display embedded cover images in audio streams
  - Add "fullscreen" button in the main video UI
  - Fix for new IdolPlParser API
  - Fix the time elapsed not showing up in the status bar
    with newer GTK+
  - Fix translation of skip menu items in RTL languages
  - Make "Esc" passing the keyboard focus to the video canvas
  - Close RTSP streams, and clean up temporary buffer files on exit

* Web browser plugin:
  - Fix a number of Javascript functions not working in the QuickTime plugin
  - Implement SetURL for QuickTime

* Plugins:
  - Remove built-in DVB support, and rely on mate-dvb-daemon instead
  - Convert D-Bus plugin to MPRIS plugin

* GStreamer backend:
  - Fix playback of files from archives
  - Make track switching faster

2.29.2 (changes since latest stable 2.28.4 release):
* General:
  - Add disk buffering for QuickTime and Flash video files,
    and only start playing the file when it can be played
    uninterrupted until the end
  - Add support for authentication for HTTP and RTSP streams
  - Make track switching faster
  - Don't save the volume anymore, the sound system should
  - Use cairo to draw the logo

* Movie Player:
  - Add OSD when keyboard or remote keys are pressed
  - Add support for reverse frame-stepping
  - Add support for playing back DVDs and VCDs from mounted
    ISO images
  - Include the movie name when taking screenshots
  - Use gallery instead of screenshot in suggested gallery file names

* Browser plugin:
  - Use referrer information
  - Add support for subtitles in the VideLAN API
  - Remove "Complex" (RealMedia compatible) plugin

2.28.0:
* Use name from the playlist for the recent item
* Set the stream volume using PulseAudio
* Fix a crash in the web browser plugin with WebKit
* Fix playback of YouTube videos
* Fix possible hangs after playing a remote file

2.27.92:
* Movie Player:
  - Remember the current file position when Pause is clicked
  - Show menu item description in status bar
  - Make Home and End keys seek to the start and the end of streams
  - Use Idol's icon as the main window logo
  - Make screenshot dialogue more like MATE's
  - Bug fixes:
    - Fix crash when a storage volume is encrypted
    - Make shuffle playlists faster with a large number of items
    - Fix the artist not showing up in the window title
    - Detect non-xine DVB channel configurations
    - Select the directory in which the movie is by default,
      when looking for a subtitle

* Web browser plugin:
  - Use the user-agents for the plugins we're mimicking
  - Show controls for the Cone plugin when VLC compatibility
    isn't explicitely asked

* Thumbnailer:
  - Make it possible to output thumbnails without the film borders

* Plugins:
  - Add plugin authoring tutorial

2.27.2:
* Movie Player:
  - Add frame-by-frame stepping
  - Better fallback names for audio and languages tracks
  - Make the arrow keys navigate DVD menus when one is loaded

src/idol.c
/* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*- */
/* 
 * Copyright (C) 2001-2007 Bastien Nocera <hadess@hadess.net>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA.
 *
 * The Idol project hereby grant permission for non-gpl compatible GStreamer
 * plugins to be used and distributed together with GStreamer and Idol. This
 * permission are above and beyond the permissions granted by the GPL license
 * Idol is covered by.
 *
 * Monday 7th February 2005: Christian Schaller: Add exception clause.
 * See license_change file for details.
 *
 */

#include "config.h"

#include <glib-object.h>
#include <glib/gi18n.h>
#include <gtk/gtk.h>
#include <string.h>

#ifdef GDK_WINDOWING_X11
/* X11 headers */
#include <gdk/gdkx.h>
#include <X11/Xlib.h>
#endif

#include "idol.h"
#include "idol-private.h"
#include "idol-interface.h"
#include "idol-options.h"
#include "idol-menu.h"
#include "idol-session.h"
#include "idol-uri.h"
#include "idol-preferences.h"
#include "idol-sidebar.h"
#include "video-utils.h"

static void
long_action (void)
{
	while (gtk_events_pending ())
		gtk_main_iteration ();
}

static UniqueResponse
idol_message_received_cb (UniqueApp         *app,
			   int                command,
			   UniqueMessageData *message_data,
			   guint              time_,
			   Idol             *idol)
{
	char *url;

	if (message_data != NULL)
		url = unique_message_data_get_text (message_data);
	else
		url = NULL;

	idol_action_remote (idol, command, url);

	g_free (url);

	return UNIQUE_RESPONSE_OK;
}

static void
about_url_hook (GtkAboutDialog *about,
	        const char *link,
	        gpointer user_data)
{
	GError *error = NULL;

	if (!gtk_show_uri (gtk_widget_get_screen (GTK_WIDGET (about)),
	                   link,
	                   gtk_get_current_event_time (),
	                   &error))
	{
	        idol_interface_error (_("Could not open link"),
	                               error->message,
	                               GTK_WINDOW (about));
	        g_error_free (error);
	}
}


static void
about_email_hook (GtkAboutDialog *about,
		  const char *email_address,
		  gpointer user_data)
{
	char *escaped, *uri;

	escaped = g_uri_escape_string (email_address, NULL, FALSE);
	uri = g_strdup_printf ("mailto:%s", escaped);
	g_free (escaped);

	about_url_hook (about, uri, user_data);
	g_free (uri);
}

/* Debug log message handler: discards debug messages unless Idol is run with IDOL_DEBUG=1.
 * If we're building in the source tree, enable debug messages by default. */
static void
debug_handler (const char *log_domain,
               GLogLevelFlags log_level,
               const char *message,
               MateConfClient *gc)
{
	static int debug = -1;

	if (debug < 0)
		debug = mateconf_client_get_bool (gc, MATECONF_PREFIX"/debug", NULL);

	if (debug)
		g_log_default_handler (log_domain, log_level, message, NULL);
}

int
main (int argc, char **argv)
{
	Idol *idol;
	MateConfClient *gc;
	GError *error = NULL;
	GOptionContext *context;
	GOptionGroup *baconoptiongroup;
	char *sidebar_pageid;

	bindtextdomain (GETTEXT_PACKAGE, MATELOCALEDIR);
	bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
	textdomain (GETTEXT_PACKAGE);

#ifdef GDK_WINDOWING_X11
	if (XInitThreads () == 0)
	{
		gtk_init (&argc, &argv);
		g_set_application_name (_("Idol Movie Player"));
		idol_action_error_and_exit (_("Could not initialize the thread-safe libraries."), _("Verify your system installation. Idol will now exit."), NULL);
	}
#endif

	g_thread_init (NULL);
	g_type_init ();

	/* Handle command line arguments */
	context = g_option_context_new (N_("- Play movies and songs"));
	baconoptiongroup = bacon_video_widget_get_option_group();
	g_option_context_add_main_entries (context, all_options, GETTEXT_PACKAGE);
	g_option_context_set_translation_domain(context, GETTEXT_PACKAGE);
	g_option_context_add_group (context, baconoptiongroup);

	g_option_context_add_group (context, gtk_get_option_group (TRUE));
	idol_session_add_options (context);
	if (g_option_context_parse (context, &argc, &argv, &error) == FALSE) {
		g_print (_("%s\nRun '%s --help' to see a full list of available command line options.\n"),
				error->message, argv[0]);
		g_error_free (error);
	        g_option_context_free (context);
		idol_action_exit (NULL);
	}
	g_option_context_free (context);

	g_set_application_name (_("Idol Movie Player"));
	gtk_window_set_default_icon_name ("idol");
	g_setenv("PULSE_PROP_media.role", "video", TRUE);
	gtk_about_dialog_set_url_hook (about_url_hook, NULL, NULL);
	gtk_about_dialog_set_email_hook (about_email_hook, NULL, NULL);

	gc = mateconf_client_get_default ();
	if (gc == NULL)
	{
		idol_action_error_and_exit (_("Idol could not initialize the configuration engine."), _("Make sure that MATE is properly installed."), NULL);
	}

	/* Debug log handling */
	g_log_set_handler (G_LOG_DOMAIN, G_LOG_LEVEL_DEBUG, (GLogFunc) debug_handler, gc);

	/* Build the main Idol object */
	idol = g_object_new (IDOL_TYPE_OBJECT, NULL);
	idol->gc = gc;

	/* IPC stuff */
	if (optionstate.notconnectexistingsession == FALSE) {
		idol->app = unique_app_new ("org.mate.Idol", NULL);
		idol_options_register_remote_commands (idol);
		if (unique_app_is_running (idol->app) != FALSE) {
			idol_options_process_for_server (idol->app, &optionstate);
			gdk_notify_startup_complete ();
			idol_action_exit (idol);
		} else {
			idol_options_process_early (idol, &optionstate);
		}
	} else {
		idol_options_process_early (idol, &optionstate);
	}

	/* Main window */
	idol->xml = idol_interface_load ("idol.ui", TRUE, NULL, idol);
	if (idol->xml == NULL)
		idol_action_exit (NULL);

	idol->win = GTK_WIDGET (gtk_builder_get_object (idol->xml, "idol_main_window"));

	/* Menubar */
	idol_ui_manager_setup (idol);

	/* The sidebar */
	playlist_widget_setup (idol);

	/* The rest of the widgets */
	idol->state = STATE_STOPPED;
	idol->seek = GTK_WIDGET (gtk_builder_get_object (idol->xml, "tmw_seek_hscale"));
	idol->seekadj = gtk_range_get_adjustment (GTK_RANGE (idol->seek));
	idol->volume = GTK_WIDGET (gtk_builder_get_object (idol->xml, "tmw_volume_button"));
	idol->statusbar = GTK_WIDGET (gtk_builder_get_object (idol->xml, "tmw_statusbar"));
	idol->seek_lock = FALSE;
	idol->fs = idol_fullscreen_new (GTK_WINDOW (idol->win));
	gtk_scale_button_set_adjustment (GTK_SCALE_BUTTON (idol->fs->volume),
					 gtk_scale_button_get_adjustment (GTK_SCALE_BUTTON (idol->volume)));
	gtk_range_set_adjustment (GTK_RANGE (idol->fs->seek), idol->seekadj);

	idol_session_setup (idol, argv);
	idol_setup_file_monitoring (idol);
	idol_setup_file_filters ();
	idol_setup_play_disc (idol);
	idol_callback_connect (idol);
	sidebar_pageid = idol_setup_window (idol);

	/* Show ! gtk_main_iteration trickery to show all the widgets
	 * we have so far */
	if (optionstate.fullscreen == FALSE) {
		gtk_widget_show (idol->win);
		idol_gdk_window_set_waiting_cursor (gtk_widget_get_window (idol->win));
		long_action ();
	} else {
		gtk_widget_realize (idol->win);
	}

	idol->controls_visibility = IDOL_CONTROLS_UNDEFINED;

	/* Show ! (again) the video widget this time. */
	video_widget_create (idol);
	gtk_widget_grab_focus (GTK_WIDGET (idol->bvw));
	idol_fullscreen_set_video_widget (idol->fs, idol->bvw);

	if (optionstate.fullscreen != FALSE) {
		gtk_widget_show (idol->win);
		gdk_flush ();
		idol_action_fullscreen (idol, TRUE);
	}

	/* The prefs after the video widget is connected */
	idol_setup_preferences (idol);

	idol_setup_recent (idol);

	/* Command-line handling */
	idol_options_process_late (idol, &optionstate);

	/* Initialise all the plugins, and set the default page, in case
	 * it comes from a plugin */
	idol_object_plugins_init (idol);
	idol_sidebar_set_current_page (idol, sidebar_pageid, FALSE);
	g_free (sidebar_pageid);

	if (idol->session_restored != FALSE) {
		idol_session_restore (idol, optionstate.filenames);
	} else if (optionstate.filenames != NULL && idol_action_open_files (idol, optionstate.filenames)) {
		idol_action_play_pause (idol);
	} else {
		idol_action_set_mrl (idol, NULL, NULL);
	}

	/* Set the logo at the last minute so we won't try to show it before a video */
	bacon_video_widget_set_logo (idol->bvw, "idol");

	if (optionstate.fullscreen == FALSE)
		gdk_window_set_cursor (gtk_widget_get_window (idol->win), NULL);

	if (idol->app != NULL) {
		g_signal_connect (idol->app, "message-received",
				  G_CALLBACK (idol_message_received_cb), idol);
	}

	gtk_main ();

	return 0;
}
